packages/core-data/src/private-selectors.ts

```diff
@@ -105,3 +106,149 @@ export function getEntityRecordPermissions(
 export function getRegisteredPostMeta( state: State, postType: string ) {
 	return state.registeredPostMeta?.[ postType ] ?? {};
 }
+
+function normalizePageId( value: number | string | undefined ): string | null {
+	if ( ! value || ! [ 'number', 'string' ].includes( typeof value ) ) {
+		return null;
+	}
+
+	// We also need to check if it's not zero (`'0'`).
+	if ( Number( value ) === 0 ) {
+		return null;
+	}
+
+	return value.toString();
+}
+
+interface SiteData {
+	show_on_front?: string;
+	page_on_front?: string | number;
+	page_for_posts?: string | number;
+}
+
+export const getHomePage = createRegistrySelector( ( select ) =>
+	createSelector(
+		() => {
+			const siteData = select( STORE_NAME ).getEntityRecord(
+				'root',
+				'site'
+			) as SiteData | undefined;
+			if ( ! siteData ) {
+				return null;
+			}
+			const homepageId =
+				siteData?.show_on_front === 'page'
+					? normalizePageId( siteData.page_on_front )
+					: null;
+			if ( homepageId ) {
+				return { postType: 'page', postId: homepageId };
+			}
+			const frontPageTemplateId = select(
+				STORE_NAME
+			).getDefaultTemplateId( {
+				slug: 'front-page',
+			} );
+			return { postType: 'wp_template', postId: frontPageTemplateId };
+		},
+		() => [
+			select( STORE_NAME ).getEntityRecord( 'root', 'site' ),
+			select( STORE_NAME ).getDefaultTemplateId( {
+				slug: 'front-page',
+			} ),
```

Comment:
The trouble with this is that the `getDefaultTemplateId` call triggers the resolver and a `wp/v2/templates/lookup` REST request. But the actual selector doesn't always call `getDefaultTemplateId`. If there is a valid homepage, it returns early and the REST request wasn't needed.

Because the `select` is same-store we can solve this by calling the plain selector function in the get-dependants callback. That's exactly what we need: just read the data from `state` and don't do any side effects.

I don't know how we would solve this if the `select` was cross-store. Interesting problem for @ellatrix or @Mamaduka who have plenty of experience working with complex selectors: do we have any existing prior art for this?

Reply:
you're right, I updated here to use the selectors directly, I think that's enough.